dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp

```diff
@@ -1652,15 +1716,80 @@ generateF90InterfaceSI(FortranInterfaceModuleGen& fimGen,
+    verticalbound_names.push_back(stencilInstantiation->getMetaData().getNameFromAccessID(fieldID) +
+                                  "_kvert_max");
+  }
+
   // memory management functions for production interface
   FortranInterfaceAPI setup("setup_" + stencilInstantiation->getName());
   FortranInterfaceAPI free("free_" + stencilInstantiation->getName());
   setup.addArg("mesh", FortranAPI::InterfaceType::OBJ);
   setup.addArg("k_size", FortranAPI::InterfaceType::INTEGER);
   setup.addArg("stream", FortranAPI::InterfaceType::CUDA_STREAM_T);
+  for(auto fieldID : getUsedFields(stencil, {dawn::iir::Field::IntendKind::Output,
+                                             dawn::iir::Field::IntendKind::InputOutput})) {
+    setup.addArg(stencilInstantiation->getMetaData().getNameFromAccessID(fieldID) + "_kmax",
+                 FortranAPI::InterfaceType::INTEGER);
+  }
 
   fimGen.addInterfaceAPI(std::move(setup));
   fimGen.addInterfaceAPI(std::move(free));
+
+  FortranWrapperAPI setupWrapper("wrap_setup_" + stencilInstantiation->getName());
+  setupWrapper.addArg("mesh", FortranAPI::InterfaceType::OBJ);
+  setupWrapper.addArg("k_size", FortranAPI::InterfaceType::INTEGER);
+  setupWrapper.addArg("stream", FortranAPI::InterfaceType::CUDA_STREAM_T);
+
+  for(auto fieldID : getUsedFields(stencil, {dawn::iir::Field::IntendKind::Output,
+                                             dawn::iir::Field::IntendKind::InputOutput})) {
+    setupWrapper.addOptArg(stencilInstantiation->getMetaData().getNameFromAccessID(fieldID) +
+                               "_kmax",
+                           FortranAPI::InterfaceType::INTEGER);
+  }
+
+  setupWrapper.addBodyLine("");
+
+  for(int i = 0; i < verticalbound_names.size(); i++) {
+    setupWrapper.addBodyLine("integer(c_int) :: " + verticalbound_names[i]);
+  }
+
+  setupWrapper.addBodyLine("");
+
+  for(auto fieldID : getUsedFields(stencil, {dawn::iir::Field::IntendKind::Output,
+                                             dawn::iir::Field::IntendKind::InputOutput})) {
+    setupWrapper.addBodyLine("if (present(" +
+                             stencilInstantiation->getMetaData().getNameFromAccessID(fieldID) +
+                             "_kmax" + ")) then");
+    setupWrapper.addBodyLine(
+        "  " + stencilInstantiation->getMetaData().getNameFromAccessID(fieldID) + "_kvert_max" +
+        " = " + stencilInstantiation->getMetaData().getNameFromAccessID(fieldID) + "_kmax");
+    setupWrapper.addBodyLine("else");
+    setupWrapper.addBodyLine("  " +
+                             stencilInstantiation->getMetaData().getNameFromAccessID(fieldID) +
+                             "_kvert_max" + " = k_size");
+    setupWrapper.addBodyLine("endif");
+    setupWrapper.addBodyLine("");
+  }
+
+  setupWrapper.addBodyLine("call setup_" + stencilInstantiation->getName() + " &");
+
+  setupWrapper.addBodyLine("( &");
+  setupWrapper.addBodyLine("   mesh, &");
```

Comment:
If we were to keep this I'd suggest we either have an auto format hook somewhere in the build pipeline or an indentation stack kinda deal in the FORTRAN interface. For now I'd just extract the indent in a variable like `const std::string indent = "     "`, then subsequently use that